### ISSUE TITLE
error[E0275]: overflow evaluating the requirement `_: std::io::Read`

### ISSUE THREAD

**Issue by zhaofengli** [rust-lang/rust#57854]
This *might* be a duplicate of #39959. I'm posting a new issue here as the mechanism that the error is triggered significantly differs from that in the aforementioned issue.

## Sample code

```rust
use std::io::{Read, Write};
use std::fs::File;
use std::path::Path;

// Comment this out... and everything will work fine
use thisdoesntexistyolo::haha;

struct Something<T: Read + Write> {
    thing: T,
}

impl<'a, T: 'a + Read + Write> Something<T>
    where &'a T: Read
{
    fn new(thing: T) -> Something<T> {
        Something { thing }
    }
}

fn main() {
    let file = File::open(Path::new("doesntmatter.txt")).unwrap();
    let mut st: Something<File> = Something::new(file);
}

```

([Playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=69a4b74228bcd717f8491debc3c23318))

```
error[E0275]: overflow evaluating the requirement `_: std::io::Read`
  --> src/main.rs:22:35
   |
22 |     let mut st: Something<File> = Something::new(file);
   |                                   ^^^^^^^^^^^^^^
   |
   = help: consider adding a `#![recursion_limit="128"]` attribute to your crate
   = note: required because of the requirements on the impl of `std::io::Read` for `&tar::archive::ArchiveInner<_>`
   = note: required because of the requirements on the impl of `std::io::Read` for `&tokio_reactor::poll_evented::PollEvented<tar::archive::ArchiveInner<_>>`
   = note: required because of the requirements on the impl of `std::io::Read` for `&tokio_reactor::poll_evented::PollEvented<tokio_reactor::poll_evented::PollEvented<tar::archive::ArchiveInner<_>>>`
   = note: required because of the requirements on the impl of `std::io::Read` for `&tokio_reactor::poll_evented::PollEvented<tokio_reactor::poll_evented::PollEvented<tokio_reactor::poll_evented::PollEvented<tar::archive::ArchiveInner<_>>>>`
   = note: required because of the requirements on the impl of `std::io::Read` for `&tokio_reactor::poll_evented::PollEvented<tokio_reactor::poll_evented::PollEvented<tokio_reactor::poll_evented::PollEvented<tokio_reactor::poll_evented::PollEvented<tar::archive::ArchiveInner<_>>>>>`
   = note: required because of the requirements on the impl of `std::io::Read` for `&tokio_reactor::poll_evented::PollEvented<tokio_reactor::poll_evented::PollEvented<tokio_reactor::poll_evented::PollEvented<tokio_reactor::poll_evented::PollEvented<tokio_reactor::poll_evented::PollEvented<tar::archive::ArchiveInner<_>>>>>>`
[...]
```

## Details

- This issue will still occur if the imported package is real.
- When `use thisdoesntexistyolo::haha;` is replaced with anything more than 2 levels deep like `use thisdoesntexistyolo::a::b;`, the code will compile without the error.

**Comment by csmoe** [comment]
cc @shepmaster it seems a playground issue, cannot reproduce on a local machine.

**Comment by zhaofengli** [comment]
I can reproduce it locally with [Arch Linux](https://www.archlinux.org/packages/community/x86_64/rust/)'s rustc: 
```
rustc 1.32.0
binary: rustc
commit-hash: unknown
commit-date: unknown
host: x86_64-unknown-linux-gnu
release: 1.32.0
LLVM version: 7.0
```

I wouldn't have used Playground if I weren't having issue on a local machine. However, I did some testing with other versions, and here are some interesting results:

- :-1: `arch rustc 1.32.0` reproduced
- :-1: `playground rustc 1.32.0` reproduced
- :+1: `rustup rustc 1.32.0 (9fda7c223 2019-01-16)` cannot reproduce
- :+1: `rustup rustc 1.24.0 (4d90ac38c 2018-02-12)` cannot reproduce
- :+1: `rustup rustc 1.33.0-nightly (4c2be9c97 2019-01-22)` cannot reproduce

Looking at Arch Linux's [build script](https://git.archlinux.org/svntogit/community.git/tree/trunk/PKGBUILD?h=packages/rust) for Rust, it appears to be a normal build with a [custom config.toml](https://git.archlinux.org/svntogit/community.git/tree/trunk/config.toml?h=packages/rust).

**Comment by shepmaster** [comment]
Yes, adding just Tokio triggers this. I've updated the original comment.

**Comment by davidtwco** [comment]
> /cc @davidtwco does this feel like anything around the external crate trait suggestions you worked on?

I don't think so - the extern crate trait suggestions work (#55613) I've done hasn't landed (I feel like there was something like that I had done which landed, but I can't find it, so I suppose not).

**Comment by shepmaster** [comment]
> hasn't landed

Oh, heh, then pretty unlikely to be related 😊! Sorry to drag you in!

**Comment by shepmaster** [comment]
> but does not necessarily require a non-existent import

Oh, you said that in your original comment; sorry.

**Comment by hcpl** [comment]
I tried to minimize from my codebase while applying @zhaofengli's ideas, and came to this:

```toml
[package]
name = "example"
version = "0.1.0"
authors = ["Author"]
edition = "2018"

[dependencies]
tokio-reactor = "=0.1.9"
```

```rust
impl Foo {}

fn bar<R>(r: &R) {
    std::io::Read::read_exact(&mut r, &mut [])
    //std::io::copy(&mut r, unimplemented!())
    //std::io::BufReader::new(r)
}
```

Playground: https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=a1e1abd1001d712519ac5f18f999bb3a.

Reproducible on latest stable, beta and nightly with either of the three lines in the body of `foo`.

All elements in this code are required to trigger the requirement evaluation overflow:

* An unrelated impl for a non-existing type;
* A generic function;
* The parameter is a reference to a generic type;
* Use of `r` as if it's `std::io::Read`.

Hope this unblocks the issue!

**Comment by hcpl** [comment]
Minimized even more to a standalone version, reproducible with just `rustc`:

```rust
struct PollEvented<E> {
    phantom: std::marker::PhantomData<E>,
}

impl<'a, E> std::io::Read for &'a PollEvented<E>
where
    &'a E: std::io::Read,
{
    fn read(&mut self, buf: &mut [u8]) -> std::io::Result<usize> {
        unimplemented!()
    }
}

fn bar<R>(r: &R) {
    std::io::Read::read_exact(&mut r, &mut [])
    //std::io::copy(&mut r, unimplemented!())
    //std::io::BufReader::new(r)
}
```

which now looks very similar to https://github.com/rust-lang/rust/issues/39959#issuecomment-301342230.

Playground: https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=f0458c4cc719dcc848c4f3cc6242afba.

**Comment by elbaro** [comment]
Having the same problem. It does not even say which file/line is the problem.
rustc 1.35.0-nightly (82e2f3ec2 2019-03-20)

```sh
❯ cargo build
   Compiling proj v0.2.0 (../proj)
error[E0275]: overflow evaluating the requirement `&tokio_reactor::poll_evented::PollEvented<_>: std::io::Write`
  |
  = help: consider adding a `#![recursion_limit="10"]` attribute to your crate
  = note: required because of the requirements on the impl of `std::io::Write` for `&tokio_reactor::poll_evented::PollEvented<tokio_reactor::poll_evented::PollEvented<_>>`
  = note: required because of the requirements on the impl of `std::io::Write` for `&tokio_reactor::poll_evented::PollEvented<tokio_reactor::poll_evented::PollEvented<tokio_reactor::poll_evented::PollEvented<_>>>`
  = note: required because of the requirements on the impl of `std::io::Write` for `&tokio_reactor::poll_evented::PollEvented<tokio_reactor::poll_evented::PollEvented<tokio_reactor::poll_evented::PollEvented<tokio_reactor::poll_evented::PollEvented<_>>>>`
  = note: required because of the requirements on the impl of `tokio_io::async_write::AsyncWrite` for `&tokio_reactor::poll_evented::PollEvented<tokio_reactor::poll_evented::PollEvented<tokio_reactor::poll_evented::PollEvented<tokio_reactor::poll_evented::PollEvented<tokio_reactor::poll_evented::PollEvented<_>>>>>`

error: aborting due to previous error

For more information about this error, try `rustc --explain E0275`.
error: Could not compile `proj`.

To learn more, run the command again with --verbose.
```

**Comment by leshow** [comment]
I am also having this issue, on both stable (1.33) and nightly (1.35).

```bash
error[E0275]: overflow evaluating the requirement `_: std::marker::Sized`
  |
  = help: consider adding a `#![recursion_limit="128"]` attribute to your crate
  = note: required because of the requirements on the impl of `std::io::Read` for `&tokio_reactor::poll_evented::PollEvented<_>`
  = note: required because of the requirements on the impl of `std::io::Read` for `&tokio_reactor::poll_evented::PollEvented<tokio_reactor::poll_evented::PollEvented<_>>`
```


With the following:

```rust
#[derive(Debug)]
pub struct Foo<R, D> {
    stream: R,
    _marker: PhantomData<D>,
}

impl<R, D> Foo<R, D>
where
    R: AsyncRead,
    D: DeserializeOwned,
{
    pub fn new(stream: R) -> Self {
        Self {
            stream,
            _marker: PhantomData,
        }
    }
}

impl<R, D> Future for Foo<R, D>
where
    R: AsyncRead,
    D: DeserializeOwned,
{
...
}
```

As soon as I add the `R` type parameter, rustc will fail.

I have tokio as a dependency:
```
[dependencies]
bytes = "0.4"
serde = "1.0"
serde_json = "1.0"
tokio-core = "0.1"
tokio-io = "0.1"
tokio-reactor = "0.1"
tokio-uds = "0.2"
tokio-codec = "0.1"
tokio = "0.1"
futures = "0.1"
```

**Comment by hcpl** [comment]
@leshow the problem is that there is no `impl<'a, T: AsyncRead> AsyncRead for &'a T` which would allow your code to work, instead there are specialized impls for a few `&'a Foo` with `Foo` being concrete types ([here](https://docs.rs/tokio/0.1.18/tokio/prelude/trait.AsyncRead.html#foreign-impls) and [here](https://docs.rs/tokio/0.1.18/tokio/prelude/trait.AsyncRead.html#implementors)). I'm pretty sure this is why `rustc` goes into infinite recursion by trying to match the parameter `R` in your code with `&PollEvented<R>` which in turn matches `&PollEvented<PollEvented<R>>` etc.

The solution is to use `&mut self.stream` instead: https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=8e9ad05ea7a4c0f473146241a47b3e4c. It works because of `impl<'a, T> AsyncRead for &'a mut T where T: AsyncRead + ?Sized`.

**Comment by estebank** [comment]
[The original report no longer emits E0275](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=69a4b74228bcd717f8491debc3c23318), it only complains about the invalid `use`, as it should:

```
error[E0432]: unresolved import `thisdoesntexistyolo`
 --> src/main.rs:6:5
  |
6 | use thisdoesntexistyolo::haha;
  |     ^^^^^^^^^^^^^^^^^^^ use of undeclared crate or module `thisdoesntexistyolo`
```

[The minimized report now looks like this](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=03ada12ca85809b28187973d38a2b087):

```
error[E0412]: cannot find type `Foo` in this scope
 --> src/lib.rs:1:6
  |
1 | impl Foo {}
  |      ^^^ not found in this scope

error[E0277]: the trait bound `&R: std::io::Read` is not satisfied
   --> src/lib.rs:4:13
    |
4   |     let _ = std::io::Read::read_exact(&mut r, &mut []);
    |             ^^^^^^^^^^^^^^^^^^^^^^^^^ the trait `std::io::Read` is not implemented for `&R`
    |
note: required by `read_exact`
help: consider introducing a `where` bound, but there might be an alternative better way to express this requirement
    |
3   | fn bar<R>(r: &R) where &R: std::io::Read {
    |                  ^^^^^^^^^^^^^^^^^^^^^^^

error[E0277]: the trait bound `&R: std::io::Read` is not satisfied
  --> src/lib.rs:5:13
   |
5  |     let _ = std::io::copy(&mut r, unimplemented!());
   |             ^^^^^^^^^^^^^ the trait `std::io::Read` is not implemented for `&R`
   |
help: consider introducing a `where` bound, but there might be an alternative better way to express this requirement
   |
3  | fn bar<R>(r: &R) where &R: std::io::Read {
   |                  ^^^^^^^^^^^^^^^^^^^^^^^

warning: unreachable call
 --> src/lib.rs:5:13
  |
5 |     let _ = std::io::copy(&mut r, unimplemented!());
  |             ^^^^^^^^^^^^^         ---------------- any code following this expression is unreachable
  |             |
  |             unreachable call
  |
  = note: `#[warn(unreachable_code)]` on by default

error[E0277]: the trait bound `&R: std::io::Read` is not satisfied
  --> src/lib.rs:6:37
   |
6  |     let _ = std::io::BufReader::new(r);
   |                                     ^ the trait `std::io::Read` is not implemented for `&R`
   |
note: required by `BufReader::<R>::new`
help: consider introducing a `where` bound, but there might be an alternative better way to express this requirement
   |
3  | fn bar<R>(r: &R) where &R: std::io::Read {
   |                  ^^^^^^^^^^^^^^^^^^^^^^^
```

If we apply the suggestions naively, [we get further errors](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=cd678f2897ad415ddfe21b0ef60dd812):

```
error[E0637]: `&` without an explicit lifetime name cannot be used here
 --> src/lib.rs:3:24
  |
3 | fn bar<R>(r: &R) where &R: std::io::Read {
  |                        ^ explicit lifetime name needed here

error[E0310]: the parameter type `R` may not live long enough
 --> src/lib.rs:3:28
  |
3 | fn bar<R>(r: &R) where &R: std::io::Read {
  |        -                   ^^^^^^^^^^^^^ ...so that the reference type `&'static R` does not outlive the data it points at
  |        |
  |        help: consider adding an explicit lifetime bound...: `R: 'static`
```

But if we fix those, [we get](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=5f2df54029cd36353927e4afa289f1cb) (which I believe is reasonable):

```
error[E0412]: cannot find type `Foo` in this scope
 --> src/lib.rs:1:6
  |
1 | impl Foo {}
  |      ^^^ not found in this scope

warning: unreachable call
 --> src/lib.rs:5:13
  |
5 |     let _ = std::io::copy(&mut r, unimplemented!());
  |             ^^^^^^^^^^^^^         ---------------- any code following this expression is unreachable
  |             |
  |             unreachable call
  |
  = note: `#[warn(unreachable_code)]` on by default

error[E0283]: type annotations needed
  --> src/lib.rs:5:13
   |
5  |     let _ = std::io::copy(&mut r, unimplemented!());
   |             ^^^^^^^^^^^^^ cannot infer type for type parameter `W` declared on the function `copy`
   |
   = note: cannot satisfy `_: std::io::Write`
help: consider specifying the type arguments in the function call
   |
5  |     let _ = std::io::copy::<R, W>(&mut r, unimplemented!());
   |                          ^^^^^^^^
```
